File: apps/java/design-pattern/src/main/java/pattern/proxy/AccountProtectionProxy.java
package pattern.proxy;

/**
 * 保護プロキシ。
 * 所有者のみがアクセスできるよう制御する。
 */
public class AccountProtectionProxy implements BankAccount {
    private final BankAccount realAccount;
    private final String ownerName;

    public AccountProtectionProxy(BankAccount realAccount, String ownerName) {
        this.realAccount = realAccount;
        this.ownerName = ownerName;
    }

    @Override
    public void deposit(int amount) {
        checkAccess();
        realAccount.deposit(amount);
    }

    @Override
    public void withdraw(int amount) {
        checkAccess();
        realAccount.withdraw(amount);
    }

    @Override
    public int getBalance() {
        checkAccess();
        return realAccount.getBalance();
    }

    private void checkAccess() {
        String currentUser = System.getProperty("user.name");
        if (!ownerName.equals(currentUser)) {
            throw new IllegalAccessError(
                    "Illegal access: " + currentUser + " cannot access account.");
        }
    }
}

File: apps/ruby/design-pattern/11-proxy/lib/account_protection_proxy.rb
class AccountProtectionProxy
  def initialize(real_account, owner_name)
    @subject = real_account
    @owner_name = owner_name
  end

  def method_missing(name, *args, &block)
    check_access
    @subject.public_send(name, *args, &block)
  end

  def respond_to_missing?(name, include_private = false)
    @subject.respond_to?(name, include_private) || super
  end

  private

  def check_access
    return if Etc.getlogin == @owner_name

    raise "Illegal access: #{Etc.getlogin} cannot access account."
  end
end
